File: Binary Search Tree/530. Minimum Absolute Difference in BST.cpp
/*
530. Minimum Absolute Difference in BST

Given a binary search tree with non-negative values, find the minimum absolute difference between values of any two nodes.

Example:

Input:

   1
    \
     3
    /
   2

Output:
1

Explanation:
The minimum absolute difference is 1, which is the difference between 2 and 1 (or between 2 and 3).
Note: There are at least two nodes in this BST.



*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        long minmin = INT_MAX;
        helper(root, numeric_limits<int>::min(),numeric_limits<int>::max(), minmin);
        return minmin;
    }
    
    void helper(TreeNode* root, long maxval, long minval, long &minmin){
        if(root){
            //cout<<root->val<<" min "<<minval<<" max "<<maxval<<endl;
            minmin = min(abs(root->val - maxval),minmin);
            minmin = min(abs(root->val - minval),minmin);
            //cout<<" abs(root->val - minval) "<<abs(root->val - minval)<<endl;
            helper(root->left, root->val, minval, minmin);
            helper(root->right, maxval, root->val,minmin);
        }
    }
};


/*

Inorder Traversal: 带着prev
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        int prev = numeric_limits<int>::min();
        int result = numeric_limits<int>::max();
        dfs(root, prev, result);
        return result;
    }

private:
    void dfs(TreeNode* root, int& prev, int& result) {
        if(!root) return;
        dfs(root->left, prev, result);
        if(prev!=numeric_limits<int>::min())
            result = min(result, root->val - prev);
        prev = root->val;
        dfs(root->right, prev, result);
    }
};

/*
Inorder Traversal: 带着prev的root
*/
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        TreeNode *prev = nullptr;
        int result = numeric_limits<int>::max();
        dfs(root, prev, result);
        return result;
    }

private:
    void dfs(TreeNode* root, TreeNode* &prev, int& result) {
        if(!root) return;
        dfs(root->left, prev, result);
        if(prev!=nullptr)
            result = min(result, root->val - prev->val);
       
        prev = root;
        dfs(root->right, prev, result);
    }
};

File: Binary Search Tree/530. Minimum Absolute Difference in BST.py
"""
530. Minimum Absolute Difference in BST
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.prev, self.res = float('-infinity'), float('infinity')
        def inorder(root):
            if root:
                inorder(root.left)
                if self.prev!=float('-infinity'):
                    self.res = min(self.res, root.val - self.prev)
                self.prev = root.val
                inorder(root.right)
        inorder(root)
        return self.res


class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        dic, prev, res = [], float('-infinity'),float('infinity')
        while dic or root:
            if root:
                dic.append(root)
                root = root.left
            else:
                root = dic.pop()
                print(root.val, prev)
                if prev != float('infinity'):
                    res = min(res, root.val - prev)
                prev = root.val
                root = root.right
        return res

        

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        array = []
        def dfs(root):
            if root.left:
                dfs(root.left)
            array.append(root.val)
            if root.right:
                dfs(root.right)
        dfs(root)
        min_diff = min([array[i]-array[i-1] for i in range(1, len(array))])
        return min_diff

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        stack = []
        current = root
        res = []
        while current or stack:
            while current:
                stack.append(current)
                current = current.left
            current = stack.pop()
            res.append(current.val)
            current = current.right
            
        ans = min([res[i]-res[i-1] for i in range(1,len(res))])

        return ans
            